test/e2e/keycloak_client_test.go

```diff
@@ -192,3 +223,145 @@ func waitForClientRoles(t *testing.T, framework *test.Framework, keycloakCR keyc
+			ID:   roleID,
+			Name: roleName,
+		})
+	}
+	scopeMappings.ClientMappings[secondClientName] = secondClient
+	return &scopeMappings, nil
+}
+
+func keycloakClientScopeMappingsTest(t *testing.T, framework *test.Framework, ctx *test.Context, namespace string) error {
+	err := WaitForClientToBeReady(t, framework, namespace, testSecondKeycloakClientCRName)
+	if err != nil {
+		return err
+	}
+	keycloakCR := getDeployedKeycloakCR(framework, namespace)
+	authenticatedClient, err := MakeAuthenticatedClient(keycloakCR)
+	if err != nil {
+		return err
+	}
+
+	// create initial client with scope mappings
+	client, err := getKeycloakClientWithScopeMappingsCR(
+		namespace,
+		authenticatedClient,
+		[]string{"realmRoleA", "realmRoleB"},
+		[]string{"a", "b"})
+	if err != nil {
+		return err
+	}
+	err = Create(framework, client, ctx)
+	if err != nil {
+		return err
+	}
+	err = WaitForClientToBeReady(t, framework, namespace, testKeycloakClientCRName)
+	if err != nil {
+		return err
+	}
+
+	// add non-existent roles
+	var retrievedClient keycloakv1alpha1.KeycloakClient
+	err = GetNamespacedObject(framework, namespace, testKeycloakClientCRName, &retrievedClient)
+	if err != nil {
+		return err
+	}
+	mappings, err := getKeycloakClientScopeMappings(
+		authenticatedClient,
+		[]string{"realmRoleB", "realmRoleC", "nonexistent"},
+		[]string{"b", "c", "nonexistent"},
+	)
+	if err != nil {
+		return err
+	}
+	retrievedClient.Spec.ScopeMappings = mappings
+	err = Update(framework, &retrievedClient)
+	if err != nil {
+		return err
+	}
+	err = WaitForClientToBeFailing(t, framework, namespace, testKeycloakClientCRName)
+	if err != nil {
+		return err
+	}
+
+	// update client: delete/leave/create mappings
+	err = GetNamespacedObject(framework, namespace, testKeycloakClientCRName, &retrievedClient)
+	if err != nil {
+		return err
+	}
+	mappings, err = getKeycloakClientScopeMappings(authenticatedClient, []string{"realmRoleB", "realmRoleC"}, []string{"b", "c"})
+	if err != nil {
+		return err
+	}
+	retrievedClient.Spec.ScopeMappings = mappings
+	err = Update(framework, &retrievedClient)
+	if err != nil {
+		return err
+	}
+	err = WaitForClientToBeReady(t, framework, namespace, testKeycloakClientCRName)
+	if err != nil {
+		return err
+	}
+
+	retrievedMappings, err := authenticatedClient.ListScopeMappings(clientName, realmName)
+	if err != nil {
+		return err
+	}
+	expected := retrievedClient.Spec.ScopeMappings
+
+	difference, intersection := keycloakclient.RoleDifferenceIntersection(
+		retrievedMappings.RealmMappings,
+		expected.RealmMappings)
+	assert.Equal(t, 0, len(difference))
+	assert.Equal(t, 2, len(intersection))
+
+	difference, intersection = keycloakclient.RoleDifferenceIntersection(
+		retrievedMappings.ClientMappings[secondClientName].Mappings,
+		expected.ClientMappings[secondClientName].Mappings)
+	assert.Equal(t, 0, len(difference))
+	assert.Equal(t, 2, len(intersection))
```

Comment:
@slaskawi I changed the visibility of `RoleDifferenceIntersection`to reuse it in the test.

Reply:
Honestly, I don't like this. It make a weird dependency between the `e2e` package and the reconciler. A dependency that should never be there. In the `model` we have a `util.go` - maybe `RoleDifferenceIntersection` should go there?

Reply:
good idea, done